sdks/java/harness/src/test/java/org/apache/beam/fn/harness/FnApiDoFnRunnerTest.java

```diff
@@ -133,3333 +138,4337 @@
-    List<ThrowingRunnable> teardownFunctions = new ArrayList<>();
-
-    new FnApiDoFnRunner.Factory<>()
-        .createRunnerForPTransform(
-            PipelineOptionsFactory.create(),
-            null /* beamFnDataClient */,
-            fakeClient,
-            null /* beamFnTimerClient */,
-            TEST_TRANSFORM_ID,
-            pTransform,
-            Suppliers.ofInstance("57L")::get,
-            pProto.getComponents().getPcollectionsMap(),
-            pProto.getComponents().getCodersMap(),
-            pProto.getComponents().getWindowingStrategiesMap(),
-            consumers,
-            startFunctionRegistry,
-            finishFunctionRegistry,
-            teardownFunctions::add,
-            null /* addProgressRequestCallback */,
-            null /* bundleSplitListener */,
-            null /* bundleFinalizer */);
-
-    assertTrue(startFunctionRegistry.getFunctions().isEmpty());
-    mainOutputValues.clear();
-
-    assertThat(consumers.keySet(), containsInAnyOrder(inputPCollectionId, outputPCollectionId));
-
-    FnDataReceiver<WindowedValue<?>> mainInput =
-        consumers.getMultiplexingConsumer(inputPCollectionId);
-    assertThat(mainInput, instanceOf(HandlesSplits.class));
-
-    mainInput.accept(
-        valueInGlobalWindow(
-            KV.of(
-                KV.of("5", KV.of(new OffsetRange(0, 5), GlobalWindow.TIMESTAMP_MIN_VALUE)), 5.0)));
-    mainInput.accept(
-        valueInGlobalWindow(
-            KV.of(
-                KV.of("2", KV.of(new OffsetRange(0, 2), GlobalWindow.TIMESTAMP_MIN_VALUE)), 2.0)));
-    assertThat(
-        mainOutputValues,
-        contains(
-            valueInGlobalWindow(
-                KV.of(
-                    KV.of("5", KV.of(new OffsetRange(0, 2), GlobalWindow.TIMESTAMP_MIN_VALUE)),
-                    2.0)),
-            valueInGlobalWindow(
-                KV.of(
-                    KV.of("2", KV.of(new OffsetRange(0, 1), GlobalWindow.TIMESTAMP_MIN_VALUE)),
-                    1.0))));
-    mainOutputValues.clear();
+      FnDataReceiver<WindowedValue<?>> mainInput =
+          consumers.getMultiplexingConsumer(inputPCollectionId);
+      IntervalWindow window1 = new IntervalWindow(new Instant(5), new Instant(10));
+      IntervalWindow window2 = new IntervalWindow(new Instant(6), new Instant(11));
+      WindowedValue<?> firstValue = valueInWindows("5", window1, window2);
+      WindowedValue<?> secondValue = valueInWindows("2", window1, window2);
+      mainInput.accept(firstValue);
+      mainInput.accept(secondValue);
+      assertThat(
+          mainOutputValues,
+          contains(
+              WindowedValue.of(
+                  KV.of(
+                      "5", KV.of(new OffsetRange(0, 5), GlobalWindow.TIMESTAMP_MIN_VALUE.plus(1))),
+                  firstValue.getTimestamp(),
+                  window1,
+                  firstValue.getPane()),
+              WindowedValue.of(
+
```

Comment:
// Java uses BigDecimal so 0.2 * 170 = 63.9...
      // BigDecimal.longValue() will round down to 63 instead of the expected 64